kubectl-volsync/cmd/migration_create.go

```diff
@@ -72,34 +108,294 @@ func validateMigrationCreate(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
-func doMigrationCreate(cmd *cobra.Command, args []string) error {
-	// Create the empty relationship
-	configDir, err := cmd.Flags().GetString("config-dir")
+func newMigrationCreate(cmd *cobra.Command) (*migrationCreate, error) {
+	mc := &migrationCreate{}
+	// build struct migrationRelationship from cmd line args
+	mr, err := newmigrationRelationship(cmd)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	mc.mr = mr
+
+	mrd := &migrationRelationshipDestination{}
+	err = parsePVCName(cmd, mrd)
+	if err != nil {
+		return nil, err
+	}
+	mc.mr.data.Destination = mrd
+
+	k8sClient, err := newClient(mrd.Cluster)
+	if err != nil {
+		return nil, err
+	}
+	mc.clientObject = k8sClient
+
+	// build struct migrationRelationshipDestination from cmd line args
+	err = newMigrationRelationshipDestination(mc, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	return mc, nil
+}
+
+func parsePVCName(cmd *cobra.Command, mrd *migrationRelationshipDestination) error {
```

Comment:
Please correct my understanding, once we parse the CLI/cmd args, the placeholder to store the values is under `migrationRelationshipDestination` that is `migrationCreate.migrationRelationship.migrationRelationshipData.migrationRelationshipDestination`.
Few members are direct members of `migrationRelationshipDestination` and few are part of `migrationRelationshipDestination.Destination`. After parsing, the entire `migrationRelationshipDestination` along with `migrationRelationshipDestination.Destination` will be built. Once we are ready with the structure then we can call `Run` just to do the actual work (not required to build `migrationRelationshipDestination.Destination` structure). 

If we push the building of  `migrationRelationshipDestination.Destination` in `Run`, then we again need to parse the CLI as we don't have any other fields to store the CLI values temporarily during `newMigrationCreate`.

With the current approach
1. parse and build entire `migrationRelationshipDestination` in `newMigrationCreate`
2. do the actual work as part of `Run`

> - If the parsing is isolated in the newMigrationCreate, you have the opportunity in your tests to manually build the migrationCreate struct via a simple assignment and not worry about the parsing logic (unless that's what you're testing).
> 
> - And if you're testing the parsing, you can build the cmd and call newMigrationCreate, then directly check the contents of the struct.

Yes, this is the approach I am following.

Reply:
The parsed flags go directly in `migrationCreate`, not in the migration relationship struct since they don't need to be preserved or used across the different sub-commands.
https://github.com/JohnStrunk/volsync/blob/cli-replication/kubectl-volsync/cmd/replication_setSource.go#L36-L38

`migrationCreate.Run()` then has the logic to take the parsed options and set the relationship structure to reflect what should be in the cluster. Then it can be `migrationRelationship.Save()`-ed and applied to the cluster.

------

In #123 (cli replication), I'm experimenting with having an `Apply()` function that belongs to the replcationRelationship struct (and is therefore common across all of its sub-commands. Apply takes the relationship and updates the cluster accordingly. This allows all the sub-commands to just have logic in their `Run()` to update the relationship struct, then call Apply as the final step. I don't yet know how this will work out, but I'm hopeful.